x-pack/plugins/infra/server/lib/alerting/log_threshold/log_threshold_executor.ts

```diff
@@ -101,17 +169,124 @@ const getESQuery = (
   // Negative assertions (things that "must not" match)
   const mustNotFilters = buildFiltersForCriteria(negativeCriteria);
 
-  const query = {
+  const rangeFilter = {
+    range: {
+      [timestampField]: {
+        gte: from,
+        lte: to,
+        format: 'epoch_millis',
+      },
+    },
+  };
+
+  return { rangeFilter, mustFilters, mustNotFilters };
+};
+
+const getGroupedESQuery = (
+  params: LogDocumentCountAlertParams,
+  sourceConfiguration: InfraSource['configuration'],
+  index: string
+): object | undefined => {
+  const { groupBy } = params;
+
+  if (!groupBy || !groupBy.length) {
+    return;
+  }
+
+  const timestampField = sourceConfiguration.fields.timestamp;
+
+  const { rangeFilter, mustFilters, mustNotFilters } = buildFiltersFromCriteria(
+    params,
+    timestampField
+  );
+
+  // For group by scenarios we'll pad the interval by 20% on the left (lte) and right (gte), this is so
+  // a wider net is cast to "capture" the groups. This is to account for scenarios where we want ascertain if
+  // there were "no documents" (less than 1 for example). In these cases we may be missing documents to build the groups
+  // and match (or not match) the criteria.
+  const interval = rangeFilter.range[timestampField].lte - rangeFilter.range[timestampField].gte;
+  const twentyPercentOfInterval = (20 / 100) * interval;
+
+  const paddedRangeFilter = {
+    range: {
+      [timestampField]: {
+        gte: rangeFilter.range[timestampField].gte - twentyPercentOfInterval,
+        lte: rangeFilter.range[timestampField].lte + twentyPercentOfInterval,
+        format: 'epoch_millis',
+      },
+    },
+  };
+
+  const aggregations = {
+    groups: {
+      composite: {
+        size: COMPOSITE_GROUP_SIZE,
+        sources: groupBy.map((field, groupIndex) => ({
+          [`group-${groupIndex}-${field}`]: {
+            terms: { field },
+          },
+        })),
+      },
+      aggregations: {
+        filtered_results: {
+          filter: {
+            bool: {
+              // Scope the inner filtering back to the unpadded range
+              must: [rangeFilter, ...mustFilters],
```

Comment:
Is there any argument against putting this into the cacheable `filter` context as well?

```suggestion
              filter: [rangeFilter, ...mustFilters],
```

Reply:
No, I think I just missed this one when I was shifting everything around.